### PR TITLE
Add swr-based emulator config hooks.

### PR DESCRIPTION
**Pull request by yuchenshi**
Progress and plan:

- [x] Add new providers and hooks
- [x] Migrate Home
- [x] Migrate AppDisconnected 
- [x] Migrate Firestore
- [x] Migrate Storage
- [ ] Migrate RTDB (and migrate databases list to `useSwr`, removing redux parts)
- [ ] Migrate Auth (keep Redux locally at least for now)
- [ ] Delete tons of boilerplate for `store/config` + more

----

Note to reviewers:

Start from `src/components/common/EmulatorConfigProvider.tsx` where new stuff is introduced. See `Home` and `AppDisconnected` to see them in action. Then move on to bigger changes like Firestore or Storage. Per-commit view on GitHub can be handy -- for example Storage changes are isolated in their own commit.

### CODE REVIEW

**Comment by yuchenshi** [comment]
Ready for a first pass, will do the rest on the list if design looks good

File: src/components/common/EmulatorConfigProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  | {
+      config?: Config | null;
+      promise: Promise<Config | null | undefined>;
+    }
+  | undefined
+>(undefined);
+
+const CONFIG_API = '/api/config';
+
+export function useOnChangePromise<T>(data: T): Promise<T> {
```
Comment (NothingEverHappens) [inline src/components/common/EmulatorConfigProvider.tsx:32]: Add a comment

File: src/components/common/EmulatorConfigProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        oldData: data,
+        deferred: makeDeferred<T>(),
+      });
+      oldDeferred.resolve(data);
+    }
+  }, [data, state, setState]);
+  return state.deferred.promise;
+}
+
+export const EmulatorConfigProvider: React.FC<{ refreshInterval: number }> = ({
```
Comment (NothingEverHappens) [inline src/components/common/EmulatorConfigProvider.tsx:52]: mention why we need to refetch the config

**Review by NothingEverHappens — APPROVED** [review]

File: src/components/Firestore/testing/FirestoreTestProviders.tsx [hunk 3]
```diff
@@ -59,28 +60,22 @@ export const FirestoreTestProviders: React.FC<RenderOptions> = React.memo(
       throw new Error('FirestoreTestProviders requires a running Emulator');
     }
     const [host, port] = hostAndPort.split(':');
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply src/components/Firestore/testing/FirestoreTestProviders.tsx:62]: This is for tests where the address is passed in as an env var (see L56 above)

**Pull request merged**